ktor-network/ktor-network-tls/jvm/src/io/ktor/network/tls/SSLEngineSocket.kt

```diff
@@ -0,0 +1,145 @@
+                loop@ while (true) {
+                    destination.clear()
+                    val result = unwrapper.readAndUnwrap(destination) { destination = it } ?: break@loop
+                    destination.flip()
+
+                    if (destination.remaining() > 0) {
+                        this.channel.writeFully(destination)
+                        this.channel.flush()
+                    }
+
+                    handleResult(result)
+                    if (result.status == SSLEngineResult.Status.CLOSED) break@loop
+                }
+            } catch (cause: Throwable) {
+                error = cause
+                throw cause
+            } finally {
+                unwrapper.cancel(error)
+                engine.closeOutbound()
+                doClose(error)
+            }
+        }
+
+    @Suppress("BlockingMethodInNonBlockingContext")
+    override fun attachForWriting(channel: ByteChannel): ReaderJob =
+        reader(CoroutineName("network-tls-output"), channel) {
+            var error: Throwable? = null
+            try {
+                val source = bufferAllocator.allocateApplication(0)
+                loop@ while (true) {
+                    source.clear()
+                    if (this.channel.readAvailable(source) == -1) break@loop
+                    source.flip()
+
+                    while (source.remaining() > 0) {
+                        val result = wrapper.wrapAndWrite(source)
+
+                        handleResult(result)
+                        if (result.status == SSLEngineResult.Status.CLOSED) break@loop
+                    }
+                }
+            } catch (cause: Throwable) {
+                error = cause
+                throw cause
+            } finally {
+                engine.closeInbound()
+                doClose(error)
+            }
+        }
+
+    override fun close() {
+        engine.closeOutbound()
+        if (isActive) launch {
+            doClose(null)
+        } else {
+            if (closed.compareAndSet(expect = false, update = true)) socket.close()
+        }
+    }
+
+    private suspend fun handleResult(result: SSLEngineResult) {
+        if (result.status == SSLEngineResult.Status.CLOSED) {
+            doClose(null)
+            return
+        }
+        if (result.handshakeStatus.needHandshake) {
+            doHandshake(result.handshakeStatus)
+        }
+    }
+
+    private suspend fun doClose(cause: Throwable?) = lock.withLock {
+        if (closed.compareAndSet(expect = false, update = true)) {
+            socket.use {
+                wrapper.close(cause)
+            }
+        }
+    }
+
+    private suspend fun doHandshake(initialStatus: SSLEngineResult.HandshakeStatus) = lock.withLock {
+        var temp = bufferAllocator.allocateApplication(0)
+        var status = initialStatus
+        while (true) {
+            when (status) {
+                SSLEngineResult.HandshakeStatus.NEED_TASK -> {
+                    coroutineScope {
```

Comment:
coroutineScope will create a new job for every `NEED_TASK` status.

Reply:
we can remove this, but `delegatedTask` is potentially blocking task, and then, we should make sure, that when creating TLS socket, we don't provide Dispatchers.Default or other non IO dispatcher to it constructor.